book-content/chapters/16-the-utils-folder.md

```diff
@@ -0,0 +1,1527 @@
+  return (
+    typeof input === "object" &&
+    input !== null &&
+    "id" in input &&
+    "title" in input &&
+    "artist" in input &&
+    "year" in input
+  );
+}
+
+// ---cut---
+const run = (maybeAlbum: unknown) => {
+  if (isAlbum(maybeAlbum)) {
+    maybeAlbum.name.toUpperCase();
+  }
+};
+```
+
+To fix this, we'd need to add even more checks to `isAlbum` to ensure we're checking the types of all the properties:
+
+```typescript
+function isAlbum(input: unknown) {
+  return (
+    typeof input === "object" &&
+    input !== null &&
+    "id" in input &&
+    "title" in input &&
+    "artist" in input &&
+    "year" in input &&
+    typeof input.id === "number" &&
+    typeof input.title === "string" &&
+    typeof input.artist === "string" &&
+    typeof input.year === "number"
+  );
+}
+```
+
+But at this point, something frustrating happens - TypeScript _stops_ inferring the return value of the function. We can see this by hovering over `isAlbum`:
+
+```typescript
+// hovering over isAlbum shows:
+function isAlbum(input: unknown): boolean;
+```
+
+This is because TypeScript's type predicate inference has limits - it can only process a certain level of complexity.
+
+Not only that, but our code is now _extremely_ defensive. We're checking the existence _and_ type of every property. This is a lot of boilerplate, and might not be necessary. In fact, code like this should probably be encapsulated in a library like [Zod](https://zod.dev/).
+
+### Writing Your Own Type Predicates
+
+To solve this, we can manually annotate our `isAlbum` function with a type predicate:
+
+```typescript
+function isAlbum(input: unknown): input is Album {
+  return (
+    typeof input === "object" &&
+    input !== null &&
+    "id" in input &&
+    "title" in input &&
+    "artist" in input &&
+    "year" in input
+  );
+}
+```
+
+This annotation tells TypeScript that when `isAlbum` returns `true`, the type of the value has been narrowed to `Album`.
+
+Now, when we use `isAlbum`, TypeScript will infer it correctly:
+
+```typescript
+const run = (maybeAlbum: unknown) => {
+  if (isAlbum(maybeAlbum)) {
+    maybeAlbum.name.toUpperCase(); // No error!
+  }
+};
+```
+
+This can ensure that you get the same type behavior from complex type guards.
+
+### Type Predicates Can be Unsafe
+
+Authoring your own type predicates can be a little dangerous. TypeScript doesn't track if the type predicate's runtime behavior matches the type predicate's type signature.
+
+```typescript
+function isNumber(input: unknown): input is number {
+  return typeof input === "string";
+}
+```
+
+In this case, TypeScript _thinks_ that `isNumber` checks if something is a number. But in fact, it checks if something is a string! There are no guarantees that the runtime behavior of the function matches the type signature.
+
+This is a common pitfall when working with type predicates - it's important to consider them about as unsafe as `as` and `!`.
+
+## Assertion Functions
```

Comment:
there is also this second type of an assertion function:
```ts
function assert(cond: boolean): asserts cond {
  if (!cond) {
    throw new Error("Failed");
  }
}
```

Notice how it uses the arbitrary boolean expression instead of some hardcoded type for the input parameter